src/System.Runtime.Numerics/src/System/Numerics/BigNumber.cs

```diff
@@ -514,11 +502,18 @@ internal static char ParseFormatSpecifier(string format, out int digits)
             return (char)0; // Custom format
         }
 
-        private static string FormatBigIntegerToHexString(BigInteger value, char format, int digits, NumberFormatInfo info)
+        private static string FormatBigIntegerToHex(BigInteger value, char format, int digits, NumberFormatInfo info)
         {
-            StringBuilder sb = new StringBuilder();
-            byte[] bits = value.ToByteArray();
-            string fmt = null;
+            Debug.Assert(format == 'x' || format == 'X');
+
+            // Get the bytes that make up the BigInteger.
+            Span<byte> bits = stackalloc byte[64]; // arbitrary limit to switch from stack to heap
+            bits = value.TryWriteBytes(bits, out int bytesWritten) ?
+                bits.Slice(0, bytesWritten) :
+                value.ToByteArray();
+
+            Span<char> stackSpace = stackalloc char[128];
```

Comment:
if bits.Length is large, we can predict that 128 won't be enough. We could tell VSB a good guess at the capacity we will need, and it will save it growing (potentially pathologically, reallocating every character)

Reply:
> potentially pathologically, reallocating every character

As I noted above, that will currently only happen after 2M characters.

> We could tell VSB a good guess at the capacity we will need

I can add a method to be called ahead of the loops below.

Reply:
> I can add a method to be called ahead of the loops below.

Actually, I already have that: AppendSpan.  I can just use it here.